### PR TITLE
Upgrade

### PR DESCRIPTION
**Pull request by henryiii**
Fixing some leaky abstractions, support for static type hints, use modern package design.

### CODE REVIEW

**Comment by henryiii** [comment]
Tomorrow or Wednesday (I have a 6 hour FPGA training tomorrow) I'll add GHA and write up a GHA guide for Scikit-HEP.org. Then we should be ready to hepunits release 1.2, the first Scikit-HEP package with static typing support!

**Comment by jpivarski** [comment]
> The first Scikit-HEP package with static typing support!

Congrats! That's a milestone and a step in a good direction.

(Is there a chance that units could _be_ types? The reason numbers with units aren't classes in hepunits is because of performance considerations, but if the unit conversion or at least checking is a type-check, then the penalty is only at startup time.)

File: src/hepunits/units/units.py [hunk 1]
```diff
@@ -63,91 +63,91 @@
 
 from math import pi
 
-from .prefixes import *
+from . import prefixes as _pre
```
Comment (eduardo-rodrigues) [inline src/hepunits/units/units.py:66]: This is what you refer to as leaky abstractions? Just for my understanding.

**Comment by eduardo-rodrigues** [comment]
> The first Scikit-HEP package with static typing support!
> 
> Congrats! That's a milestone and a step in a good direction.
> (Is there a chance that units could be types? The reason numbers with units aren't classes in hepunits is because of performance considerations, but if the unit conversion or at least checking is a type-check, then the penalty is only at startup time.)

There's quite a bit more to it. We would then have to go the way AstroPy did - then we would need numpy arrays with units, numpy ufuncs would have to work with these, etc. And what about wrapped C++ code, numba ...? I'm convinced it's not worth the effort in HEP. I would really need to see the huge gain from having a heavy system. With this simple system you can either just work "as before", without units, or use them the help readibility, for example. We do the latter in LHCb and that's no problem.

File: src/hepunits/units/units.py (hunk 1 above)
Comment (henryiii) [reply src/hepunits/units/units.py:66]: The problem is you have circular `import *`'s. MyPy can handle circular imports or starred imports, but not both. `hepunits/__init__.py` imports the subpackages, but subpackages import from a different subpackage, which imports the parent package, which causes a circle. Python is smart enough to break the loop, and MyPy is too, but not if there are too many `import *`'s. The `import *`'s should only be outward facing; internally you should not be `import *`'ing from one subpackage to another.

Comment (eduardo-rodrigues) [reply src/hepunits/units/units.py:66]: OK, thanks. Makes sense.

**Comment by henryiii** [comment]
> Is there a chance that units could be types?

Sure. There are three options:

a) Currently, by simply fixing issues like circular starred imports and adding a py.typed marker file, this supports typing with no changes (since there are no functions in the package). Types are floats and ints (we should probably be careful to make units floats, and prefixes ints). This does not support anything special with units.
b) We could add float type aliases, like Length. This would allow either floats or Length to be used and checked; mostly it just gives you slightly better error messages when running MyPy.
c) We can invent a custom type system in `.pyi` (Python interface) files. This could either be alongside the current files, or (better, probably) in a hepunits-stubs package. This would allow us to invent types that have behaviors that only exist in type checkers. (I'm not sure if -stubs packages are automatically installed or treated differently in Pip, I do know it's a special designation). If we could make it "opt-in", that would be fantastic. We could work it in slowly, and make it completely optional. Maybe we could even combine these last two approaches, and have  installing the stubs file enable the advanced type checking!

> numpy arrays with units

NumPy arrays are really in-progress currently; we'd have to wait and see what happens (`numpy-stubs` is where it's being worked on). If it's opt-in, we could work on non-array packages first (Particle). I think we could manage this later on with generics and/or aliases. Again, this is only at check-time (think compile-time), and not at runtime.

---

Either way, that shouldn't stop this from moving forward, as it allows MyPy to be applied to Particle (which, sadly, shows quite a few issues we probably won't fix until 0.11).

**Pull request merged**